Logging feature in#7

Mephi-Dross opened the pull request:
This is the first version of the logging system. For now it contains the _groundwork_ (interfaces/enums/Logger class) as well as the first _writer_ (ConsoleWriter). I've also added _tests_ for it, but I'm not sure if they're correctly set up. Please check 'em for me when you've got the time.

Commit:
chore: Add /.*/ folders to gitignore

These folders are often hidden and contain temporary or unnecessary VisualStudio files.

Commit:
Revert "chore: Add /.*/ folders to gitignore"

This reverts commit 98c2e78fca063c1e08e0383db8a643644670c6ac.

Commit:
chore: Adjust .gitignore to ignore files that start with '.'

These files and folders are often hidden and contain temporary or unnecessary VisualStudio files.

Commit:
feat: Add first version of the logging system

This intial version of the logging only contains logging to Console. However the groundwork for any number of additional loggers has been set. In addition I've added integration tests for the console logger.

texel-sensei reviewed (changes requested):
* A logger should be able to support an arbitrary amount of log writers and send the log to all of them, not just one.

* The LogType enum should be removed. Add the LogWriter by just passing a constructed object from the calling code.

* Rename the log levels to more meaningful names.
Maybe `Trace`, `Debug`, `Info`, `Warning`, `Error`. I like the `EverythingIsOnFire` one though 👍

texel-sensei commented on src/core/ayayaa-server/ayayaa-logging/Logger.cs:
```diff
@@ -0,0 +1,45 @@
+                case LogType.Remoteserver:
+                    break;
+                default:
+                    break;
```
This line is dangerous, this would create an Logger without writer. All log output would then be silently ignored and/or possible NullReferenceExceptions.

Commit:
feat(log): Extend Logger, adjust Writer, add Tests

Added formating to Logger.
Removed LogType.
Adjusted IWriter to show all functions.
Changed FormMessage to SerializeMessage.
Adjusted ExceptionHandling in all classes.
Added LoggerException for better ExceptionHandling.
Added Tests for Writers.
Adjusted Tests for Logger.

Mephi-Dross commented:
- Ability to handle multiple Writers acknowledeged, not yet implemented.
- LogType has been removed and Logger adjusted accordingly.
- Renaming of LogPriority acknowledged, not yet implemented.

This update primarily focused on changing the way the text in the logger/writer is handled as well as how error handling was adjusted. Furthermore the code has been changed to be more test friendly and additional tests have been added.

Commit:
feat(log): Adjust Logger to handle writers

Adjusted LogPriority-Names.
Adjusted Logger to handle multiple writers.
Adjusted tests to new Logger.

Mephi-Dross commented:
- LoggingPriority names have been adjusted.
- Logger can now handle multiple writers.
- Tests have been adjusted to reflect Logger changes.

texel-sensei commented on src/core/ayayaa-server/ayayaa-logging/Interfaces/ILogger.cs:
```diff
@@ -0,0 +1,17 @@
+
+        void AddWriter(IWriter writer, LogPriority priority);
+        void WriteToLog(string message, LogPriority priority, IWriter log);
+        Dictionary<IWriter, bool> WriteToLogs(string message, LogPriority priority);
```
Returning a dictionary of success/failure is probably overkill here. A single boolean to indicate if the log was written by *any* writer would be ok.

Considering that most log output will never be error checked.

texel-sensei commented on src/core/ayayaa-server/ayayaa-logging/Interfaces/IWriter.cs:
```diff
@@ -0,0 +1,30 @@
+        /// (e.g. byte[] for sending it to a remote host.)
+        /// </summary>
+        /// <param name="message"></param>
+        object SerializeMessage(string message);
```
"Internal use only" is exactly the opposite of having an interface. I think it's better to just remove those functions from the interface and keep them as implementation detail.

texel-sensei reviewed (changes requested):
The IWriter interface is still a bit bloated, if a method is commented with "Internal use only" then it should not be part of an Interface

Commit:
improvement(log): Adjust interfaces and Logger

Removed unnecessary functions from interfaces.
Removed tests that relied on these functions.
Changed visibility of certain functions to private.

Mephi-Dross commented:
I agree with removing stuff from the interfaces and making the functions private. To be honest, it was me going overboard to write testable code since I still need to improve my knowledge there.

I disagree with the Dictionary though. In 90% of cases you don't care about the fail states, but in those cases you do care it provides a better view of where the problem is at a glance. No need to search in the logs themselves if stuff actually succeeded or not, you can check instantly in the code. It's not like Dictionaries are a giant performance hog.

texel-sensei committed:
Merge branch 'develop' into logging_feature_IN#7
